### PR TITLE
[16.0][ADD] edi_account_edifact_oca

### PR DESCRIPTION
**Pull request by QuocDuong1306**
This module supports sending account invoices from EDI channels to the storage system.

Depend on:
- https://github.com/OCA/edi/pull/853
- https://github.com/OCA/edi/pull/854
- https://github.com/OCA/edi-framework/pull/26
- https://github.com/OCA/edi-framework/pull/39

### CODE REVIEW

File: edi_account_edifact_oca/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    "version": "16.0.1.0.0",
+    "development_status": "Alpha",
+    "license": "AGPL-3",
+    "author": "Camptocamp,Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/edi-framework",
+    "depends": [
+        "edi_account_oca",
+        "edi_edifact_oca",
+        "account_invoice_edifact",
+        "edi_storage_oca",
```
Comment (simahawk) [inline edi_account_edifact_oca/__manifest__.py]: This module should not depend on storage. The framework is agnostic for this reason: it does not matter where you send to or receive files from.

**Review by simahawk — CHANGES_REQUESTED** [review]
Also see my comments on https://github.com/OCA/edi/pull/854

File: edi_account_edifact_oca/components/listener.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+_logger = logging.getLogger(__name__)
+
+
+class AccountMoveEdifactListener(Component):
+    _name = "account.move.event.listener.edifact"
+    _inherit = "base.event.listener"
+    _apply_on = ["account.move"]
+
+    def on_post_account_move(self, move):
+        edifact_backend = self.env.ref("edi_edifact_oca.edi_backend_type_edifact")
```
Comment (simahawk) [inline edi_account_edifact_oca/components/listener.py]:
```suggestion
        backend_type = self.env.ref("edi_edifact_oca.edi_backend_type_edifact")
```

File: edi_account_edifact_oca/components/listener.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        if (
+            not move.partner_id
```
Comment (simahawk) [inline edi_account_edifact_oca/components/listener.py]: pls move this condition to a specific method. The direction is not needed (see the change that I proposed to the field name in the other PR).

File: edi_account_edifact_oca/demo/edi_exchange_type.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      <field name="name">Demo EDIFACT Invoice</field>
+      <field name="code">demo_EDIFACT_Invoice_out</field>
+      <field name="direction">output</field>
+      <field name="exchange_file_ext">txt</field>
+      <field name="exchange_file_auto_generate">True</field>
+      <field name="advanced_settings_edit">
+components:
+  generate:
+    usage: output.generate.account.move
+  send:
```
Comment (simahawk) [inline edi_account_edifact_oca/demo/edi_exchange_type.xml]: pls remove `send`

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by QuocDuong1306** [comment]
Hi @simahawk , I updated

File: edi_account_edifact_oca/components/generate.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+class EDIExchangeAccountMoveGenerate(Component):
+    _name = "edi.output.account.move.generate"
+    _inherit = "edi.component.output.mixin"
+    _usage = "output.generate.account.move"
+
+    def generate(self):
+        data = False
+        exchange_record = self.exchange_record
+        if exchange_record:
+            if exchange_record.model == "account.move" and exchange_record.res_id:
```
Comment (simahawk) [inline edi_account_edifact_oca/components/generate.py]: the model should be selected by `_apply_on` on the class

File: edi_account_edifact_oca/components/generate.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+class EDIExchangeAccountMoveGenerate(Component):
+    _name = "edi.output.account.move.generate"
+    _inherit = "edi.component.output.mixin"
+    _usage = "output.generate.account.move"
+
+    def generate(self):
+        data = False
+        exchange_record = self.exchange_record
+        if exchange_record:
```
Comment (simahawk) [inline edi_account_edifact_oca/components/generate.py]: I think is impossible to not have an exchange record

File: edi_account_edifact_oca/components/generate.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+            if exchange_record.model == "account.move" and exchange_record.res_id:
+                move = self.env["account.move"].browse(exchange_record.res_id)
```
Comment (simahawk) [inline edi_account_edifact_oca/components/generate.py]: there's no need to browse. Use `self.record`.

File: edi_account_edifact_oca/components/generate.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2023
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo.addons.component.core import Component
+
+
+class EDIExchangeAccountMoveGenerate(Component):
+    _name = "edi.output.account.move.generate"
```
Comment (simahawk) [inline edi_account_edifact_oca/components/generate.py]:
```suggestion
    _name = "edi.output.edifact.account.move.generate"
```

File: edi_account_edifact_oca/readme/USAGE.rst [hunk 9]
```diff
@@ -0,0 +1,5 @@
+To use this module, you need to:
+
+* Config `ID number` on the customer and the sender.
+* Config `EDI Exchange Type For Invoices` in `Invoicing` tag on the customer.
+* Config `Storage` of the EDI backend.
```
Comment (simahawk) [inline edi_account_edifact_oca/readme/USAGE.rst]: storage is not needed. Is not relevant how you send it.

File: edi_account_edifact_oca/tests/test_edi_account_edifact.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        )
+
+    def test_post_invoices(self):
+        self.invoice.action_post()
+        exchange_record = self.env["edi.exchange.record"].search(
+            [
+                ("model", "=", "account.move"),
+                ("res_id", "=", self.invoice.id),
+            ]
+        )
```
Comment (simahawk) [inline edi_account_edifact_oca/tests/test_edi_account_edifact.py]:
```suggestion
        exchange_record = self.invoice.exchange_record_ids[0]

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by QuocDuong1306** [comment]
Thank @simahawk , the PR has been updated as suggested

**Pull request closed without merging**